File: Objective-C Sample Code/Chapter 6/Top10/Top10/SessionManager.m
//
//  SessionManager.m
//  Top10
//
//  Created by Cory Bohon on 12/13/16.
//  Copyright © 2016 Cory Bohon. All rights reserved.
//

#import "SessionManager.h"
@import WatchConnectivity;

@interface SessionManager() <WCSessionDelegate>

@property (nonatomic, strong) WCSession *session;

@end

@implementation SessionManager

+ (instancetype)sharedManager
{
    static SessionManager *sharedManager = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedManager = [[self alloc] init];
        [sharedManager configureSession];
    });
    return sharedManager;
}

- (void)configureSession
{
    if (![WCSession isSupported]) {
        return;
    }
    
    self.session = [WCSession defaultSession];
    self.session.delegate = self;
    [self.session activateSession];
}

- (void)session:(WCSession *)session activationDidCompleteWithState:(WCSessionActivationState)activationState error:(NSError *)error
{
    if (error) {
        NSLog(@"Unable to activate WCSession: %@", error.localizedDescription);
    }
}

- (void)session:(WCSession *)session didReceiveFile:(WCSessionFile *)file
{
    NSURL *fileToMoveURL = [file fileURL];
    
    NSArray *filePaths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    
    NSString *path = [[filePaths firstObject] stringByAppendingPathComponent:[fileToMoveURL lastPathComponent]];
    
    NSURL *newLocationURL = [NSURL fileURLWithPath:path];
    
    __autoreleasing NSError *error = nil;
    
    [[NSFileManager defaultManager] moveItemAtURL:fileToMoveURL toURL:newLocationURL error:&error];
    
    if (error) {
        NSLog(@"Error Moving File: %@", error.localizedDescription);
    }
}

@end

File: Swift Sample Code/Chapter 6/Top10/Top 10 WatchKit App Extension/SessionManager.swift
//
//  SessionManager.swift
//  Top10
//
//  Created by Cory Bohon on 2/14/17.
//  Copyright © 2017 Cory Bohon. All rights reserved.
//

import UIKit
import WatchConnectivity

class SessionManager: NSObject, WCSessionDelegate {
    
    static let shared = SessionManager()
    var session: WCSession?
    
    override init() {
        super.init()
        self.configureSession()
    }
    
    func configureSession() {
        guard WCSession.isSupported() == true else {
            return
        }
        
        self.session = WCSession.default()
        self.session?.delegate = self
        self.session?.activate()
    }
    
    // MARK: - WCSessionDelegate 
    func session(_ session: WCSession, activationDidCompleteWith activationState: WCSessionActivationState, error: Error?) {
        
        if let error = error {
            print("Unable to activate WCSession: \(error.localizedDescription)")
        }
    }
}
